// @teambit/bvm.list/dist/gcp/gcp-version.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GcpVersion = void 0;
const version_1 = require("../version");
class GcpVersion {
    constructor(fileName, host, md5hash, timeCreated, metadata, protocol = 'https') {
        this.fileName = fileName;
        this.host = host;
        this.md5hash = md5hash;
        this.timeCreated = timeCreated;
        this.metadata = metadata;
        this.protocol = protocol;
    }
    get version() {
        return this.fileName
            .split('/')[4]
            .replace(/\.[^/.]+$/, '')
            .replace(/\.[^/.]+$/, '')
            .split('-')[1];
    }
    get url() {
        return `${this.protocol}://${this.host}/${this.fileName}`;
    }
    /** is version is stable version  */
    get stable() {
        var _a;
        if ((_a = this.metadata) === null || _a === void 0 ? void 0 : _a.stable)
            return true;
        return false;
    }
    toRemoteVersion() {
        return new version_1.RemoteVersion(this.version, this.url, this.md5hash, this.timeCreated, this.stable);
    }
    toObject() {
        return {
            url: this.url,
            version: this.version,
        };
    }
}
exports.GcpVersion = GcpVersion;
