// @teambit/bvm.commands.use/dist/use.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.command = exports.UseCmd = void 0;
const chalk_1 = __importDefault(require("chalk"));
const bvm_config_1 = require("@teambit/bvm.config");
const bvm_link_1 = require("@teambit/bvm.link");
class UseCmd {
    constructor() {
        this.aliases = ['use'];
        this.describe = 'switch used bit version';
        this.command = [
            'use [bit-version]'
        ];
    }
    builder(yargs) {
        yargs.positional('bit-version', {
            describe: 'version to use',
            type: 'string'
        })
            .example('$0 use', 'show currently used version of bit')
            .example('$0 use 1.7.5', 'switch to version 1.7.5 of bit');
        return yargs;
    }
    handler(args) {
        return __awaiter(this, void 0, void 0, function* () {
            let results;
            if (!args.bitVersion) {
                const config = bvm_config_1.Config.load();
                const currentVersion = config.getDefaultLinkVersion();
                return console.log(chalk_1.default.cyan(`currently use ${currentVersion} version`));
            }
            results = [yield (0, bvm_link_1.linkDefault)(args.bitVersion, true)];
            printOutput(results, args.verbose);
            return;
        });
    }
    ;
}
exports.UseCmd = UseCmd;
function formatOutput(linkResults, verbose = false) {
    const links = linkResults.map(result => {
        const linkTarget = verbose ? chalk_1.default.cyan(`(${result.generatedLink.target})`) : '';
        const linkSource = verbose ? chalk_1.default.cyan(`(${result.generatedLink.source})`) : '';
        return (`name ${chalk_1.default.green(result.linkName)}${linkTarget} points to version ${chalk_1.default.green(result.version)}${linkSource}`);
    });
    const summery = `successfully linked binaries`;
    return [links, summery].filter(msg => msg).join('\n');
}
function printOutput(linkResults, verbose = false) {
    const output = formatOutput(linkResults, verbose);
    console.log(output);
    return output;
}
exports.command = new UseCmd();
